pytorch_lightning/core/datamodule.py

```diff
@@ -0,0 +1,255 @@
+import inspect
+from abc import abstractmethod
+from argparse import ArgumentParser, Namespace
+from typing import Union, List, Tuple, Any
+
+from pytorch_lightning.utilities import rank_zero_warn, parsing
+from torch.utils.data import DataLoader
+
+
+class LightningDataModule(object):  # pragma: no cover
+    """
+    A DataModule standardizes the training, val, test splits, data preparation and transforms.
+    The main advantage is consistent data splits and transforms across models.
+    Example::
+        class MyDataModule(LightningDataModule):
+            def __init__(self):
+                super().__init__()
+            def prepare_data(self):
+                # download, split, etc...
+            def train_dataloader(self):
+                train_split = Dataset(...)
+                return DataLoader(train_split)
+            def val_dataloader(self):
+                val_split = Dataset(...)
+                return DataLoader(val_split)
+            def test_dataloader(self):
+                test_split = Dataset(...)
+                return DataLoader(test_split)
+    A DataModule implements 4 key methods
+    1. **prepare_data** (things to do on 1 GPU not on every GPU in distributed mode)
+    2. **train_dataloader** the training dataloader.
+    3. **val_dataloader** the val dataloader.
+    4. **test_dataloader** the test dataloader.
+    This allows you to share a full dataset without explaining what the splits, transforms or download
+    process is.
+    """
+    name: str = ...
+
+    def __init__(
+            self,
+            train_transforms=None,
+            val_transforms=None,
+            test_transforms=None,
+    ):
+        super().__init__()
+        self._train_transforms = train_transforms
+        self._val_transforms = val_transforms
+        self._test_transforms = test_transforms
+        self.dims = ()
+
+    @property
+    def train_transforms(self):
```

Comment:
Just a question: Why do you need these properties if neither setter nor getter are doing something specific like type checking? can't you just change this to public attributes?

Reply:
Going to add type checking there. does that make sense or should this still change to public attribute?

Reply:
No, with type checking it's fine :) Just in the current state it was not necessary to do it that way :)

Reply:
one advantage of setters and getters is also that you can attach documentation to them, which you wouldn't be able to if it was just a regular attribute.

Reply:
Strike my last point...when I viewed this earlier, it was on Mobile. I thought you were pointing out something else.

I'll add docs here as Adrian pointed out.